quinn-proto/src/shared.rs

```diff
@@ -181,28 +179,39 @@ impl TransportConfig {
 }
 
 /// Global configuration for the endpoint, affecting all connections
+///
+/// This type contains a private `reset_key` value which is initialized with random bytes by the
+/// `default()` constructor and can be modified using `set_reset_key()`. See the latter method
+/// for more information.
 pub struct EndpointConfig {
     /// Length of connection IDs for the endpoint.
     ///
     /// This must be either 0 or between 4 and 18 inclusive. The length of the local connection IDs
     /// constrains the amount of simultaneous connections the endpoint can maintain. The API user is
     /// responsible for making sure that the pool is large enough to cover the intended usage.
     pub local_cid_len: usize,
+    pub(crate) reset_key: [u8; 64],
```

Comment:
Why can't this be public?

Reply:
So the idea I had was that since this is private key-like material, we should maybe not make it possible to extract the key material (as it works in *ring*). But maybe this isn't so sensitive...

Reply:
I'm not sure how productive that is; the caller can always pull out the data with unsafe code, and there's a high ergonomic cost. The custom `Debug` impls that don't print it are nice, though.

Reply:
What part of the ergonomic cost do you identify as high, exactly? I did a version with `new()` that takes a key as a parameter (and fills the other fields from `default()`), but I found this version easier to explain in documentation.

We should maybe talk about the threat model, in the sense of, what happens when these keys inadvertently leak? (I understand that you could always get them out using unsafe code, but that's sort of besides the point IMO -- defense in depth matters, and this makes it harder for the secrets to escape in any case.)

Reply:
Specifically, the loss of the ability to use functional record update syntax, and the inconsistency in how the one field is written to. A large `new` function would be even worse, I agree there. I also think the added insurance against misuse is extremely marginal; users will almost never have a `Config` struct that they did not create, and when creating one they'll necessarily already have access to the raw bytes.

Reply:
Fixed this part.

If you think the leaking is not a big problem, do you really think the custom `Debug` impls are worth it compared to `derive(Debug)`?

Reply:
I don't have a strong opinion and I'm not as steeped in security idioms as I might be, but my knee-jerk reaction is that a `Debug` dump of a (value containing a) config struct is overwhelmingly the most likely channel for an accidental information leak. Perhaps some users even desire the option to stash `Debug` dumps of config into logs.

Reply:
Okay, let's keep it like this.